frameworks/FSharp/Zebra/src/App/Program.fs

```diff
@@ -4,35 +4,132 @@ open State
 open Router
 open ExecNodes
 open Middleware
-open System.Threading.Tasks
-open System.Runtime.CompilerServices
+
 open Microsoft.AspNetCore.Http
 open Microsoft.AspNetCore.Hosting
-open Microsoft.AspNetCore.Builder
-
+open Microsoft.AspNetCore
+open System
+open TemplateViewEngine
+open Npgsql
+open Dapper
+open System.Collections.Generic
 
 
 [<CLIMutable>][<Struct>] 
 type JsonStructMessage = { message : string }
 
+ [<CLIMutable>]
+type Fortune = { id: int; message: string }
+
+[<Literal>]
+let ConnectionString = "Server=tfb-database;Database=hello_world;User Id=benchmarkdbuser;Password=benchmarkdbpass;Maximum Pool Size=1024;NoResetOnClose=true;Enlist=false;Max Auto Prepare=3"
+
+
+module Simple =
+    let textFn<'T> (text:string) = 
+        let bytes =         System.Text.Encoding.UTF8.GetBytes(text)
+        let contentLength = Microsoft.Extensions.Primitives.StringValues(bytes.Length.ToString())
+        let contentType =   Microsoft.Extensions.Primitives.StringValues "text/plain"
+        fun (x:State<'T>) -> 
+            x.HttpContext.Response.Headers.["Content-Length"] <- contentLength
+            x.HttpContext.Response.Headers.["Content-Type"] <- contentType
+            let t = x.HttpContext.Response.Body.WriteAsync(bytes,0,bytes.Length)
+            let awt = t.GetAwaiter()
+            x.CurrentState <- MachineState.Complete
+            awt.OnCompleted x.Continue
+
+    let inline jsonFn<'T> (value: ^a) =
+        let bytes =         Utf8Json.JsonSerializer.Serialize< ^a>(value)
+        let contentLength = Microsoft.Extensions.Primitives.StringValues(bytes.Length.ToString())
+        let contentType =   Microsoft.Extensions.Primitives.StringValues "application/json"
+        fun (x:State<'T>) -> 
+            x.HttpContext.Response.Headers.["Content-Length"] <- contentLength
+            x.HttpContext.Response.Headers.["Content-Type"] <- contentType
+            let t = x.HttpContext.Response.Body.WriteAsync(bytes,0,bytes.Length)
+            let awt = t.GetAwaiter()
+            x.CurrentState <- MachineState.Complete
+            awt.OnCompleted x.Continue
+
+module View =
+     
+    let fortuneView =
+        html [] [
+            head [] [
+                title []  [ rawText "Fortunes" ]
+            ]        
+            body [] [
+                table [] [ 
+                    tr [] [
+                        th [] [ rawText "id" ]
+                        th [] [ rawText "message" ]
+                    ]
+                    bindFor<_,_> (fun ls -> ls :> seq<Fortune> ) (
+                        tr [] [
+                            td [] [ bindInt (fun v -> v.id) ]
+                            td [] [ bindStr (fun v -> v.message) ]
+                        ]
+                    )
+                ]
+            ]
+        ] |> compileDoc
```

Comment:
Just out of curiosity, what is happening here? I recall older implementations using a Razor template for `fortunes`, but this seems to be a simpler compiled template... thing.

Reply:
"Thing" is probably the best word to describe it alright … F# can use Razor but it's bit of a hack together of XML & C#, in F# we can us a simple DSL to represent the HTML as well as type checking all the bindings. `body`, `table` etc are all just functions that create a DSL node tree similar to XML but with the `compileDoc` it is taking in all the DSL elements and creating a processing array of either the static elements (grouped into one byte array), or a binding function that will inject values at runtime.

Reply:
That's neat.